ddtrace/internal/telemetry/metrics.py

```diff
@@ -0,0 +1,131 @@
+# -*- coding: utf-8 -*-
+import abc
+import time
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Tuple
+
+import six
+from typing_extensions import Literal
+
+from ..hostname import get_hostname
+
+
+MetricType = Literal["count", "gauge", "rate"]
+MetricTagType = Dict[str, str]
+
+
+class Metric(six.with_metaclass(abc.ABCMeta)):
+    """
+    stores metrics which will be sent to the Telemetry Intake metrics to the Datadog Instrumentation Telemetry Org
+    """
+
+    HOST_NAME = get_hostname()
+    metric_type = ""
+
+    def __init__(self, namespace, name, tags, common, interval=None):
+        # type: (str, str, MetricTagType, bool, Optional[float]) -> None
+        """
+        name: metric name
+        common: set to True if a metric is common to all tracers, false if it is python specific
+        interval: field set for gauge and rate metrics, any field set is ignored for count metrics (in secs)
+        """
+        self.name = name
+        self.common = common
+        self.interval = interval
+        self._roll_up_interval = interval
+        self.namespace = namespace
+        self._points = []  # type: List[Tuple[int, float]]
+        self._tags = tags  # type: MetricTagType
+        self._count = 0.0
+
+    @property
+    def id(self):
+        """
+        https://www.datadoghq.com/blog/the-power-of-tagged-metrics/#whats-a-metric-tag
+        """
+        return self.name + str(self._tags)
+
+    def __hash__(self):
+        return self.id
+
+    @abc.abstractmethod
+    def add_point(self, value=1.0):
+        # type: (float) -> None
+        """adds timestamped data point associated with a metric"""
+        pass
+
+    def set_tags(self, tags):
```

Comment:
It might not be particularly intuitive to a caller that this function replaces `self._tags` rather than `extend`ing it. You might either clearly document that fact or change the implementation to use `extend`.